### PR TITLE
Data model architecture [WiP]

### PR DESCRIPTION
**Pull request by sbelak**
Stubbed out functionality and removed parts that will no longer be needed for illustrative purposes

# xrays
* heuristics are written against data models
* no overloading
* no dimensions binding
* metrics optional (else use data model)
* MBQL generation is moved to the data model

**Bonus**  xrays are cheaper+faster to generate as data model is essentially a precomputation

### open questions
* should there be some generic data model?
* How do drilldown x-rays look?
* Should we retain the notion of interestingness/card score? If so where/why?

# data model
* bound during analysis/classify
* reclassify if the underlying table changes
* we'll need a cleanup process for stale/unused data model entities 
* specificity by refines relation
* model assigment: pick the most specific model among those that fit directly or via ETL
* match + instantate template

###  open questions
* **how to model relations !**
* grouping would be nice
* are string names for dimensions chill?

# ETL
* do ETL tasks always `:provides` data models or should we spec the shape of results separately 

Needs
* factor out common YAML rule file handling and language definition
* data model classifier should be in own namespace, put it in names.clj for now to illustrate similarity to current entity_type classifier 

Note this PR is based of MBTL branch for clarity

See #10071 for data model spec, and #10105 for ETL (recommended reading order: #10071 -> #10105 -> this ) 

Next steps:
1. finish off ETL (see TODO there)
2. factor out common YAML rule file handling and language definition
3. fill out the stubs in data-model.clj
4. write at least the data models for all existing entity types
5. update x-ray code to use data models
6. update x-ray definitions
7. start building cool shit

### CODE REVIEW

File: resources/data_models/orders.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  - Product
+  - Source
+  - NumOrders
+metrics:
+  TotalIncome:
+    aggregation [sum, [dimensions, Income]]
+segments:
+  ReturningUsers:
+    filter: [>, [dimension, NumOrders],  1]
+breakout_dimensions:
```
Comment (salsakran) [inline resources/data_models/orders.yaml]: minor nit, but if we're going to have two sets of dimensions in one group above `required_dimensions` and `optional_dimensions`, we should group this with them.

File: src/metabase/models/data_model.clj [hunk 2]
```diff
@@ -0,0 +1,39 @@
+(ns metabase.models.data-model
```
Comment (salsakran) [inline src/metabase/models/data_model.clj]: we should have a better naming for this so it doesn't get confused with Toucan models?

File: src/metabase/automagic_dashboards/core.clj [hunk 3]
```diff
@@ -972,6 +643,10 @@
        (remove (comp nil? second))
        (into {})))
 
+(defn- find-rule-for-entity
```
Comment (salsakran) [inline src/metabase/automagic_dashboards/core.clj]: what is a "rule" in this context? I don't think this is a good name given I have no idea what it refers to =)

File: resources/automagic_dashboards/table/TransactionTable.yaml [hunk 4]
```diff
@@ -1,87 +1,7 @@
...
-    named: category
-    max_cardinality: 10
-    score: 90
-- ProductCategoryLarge:
-    field_type: ProductTable.Category
-    named: category
-- Long: GenericTable.Longitude
-- Lat: GenericTable.Latitude
+data_model: Orders
 filters:
```
Comment (salsakran) [inline resources/automagic_dashboards/table/TransactionTable.yaml]: Micro-question -- 

what exactly is going on with this vs segments in the data model file?

eg, should the x-rays actually have any definitions of segments or common filters at all?

File: src/metabase/automagic_dashboards/core.clj [hunk 5]
```diff
@@ -653,109 +444,6 @@
...
-  [context metric]
-  (update context :metrics assoc "this" {:metric (->reference :mbql metric)
-                                         :name   (:name metric)
-                                         :score  rules/max-score}))
-
-(defmethod inject-root :default
-  [context _]
-  context)
-
-(s/defn ^:private make-context
```
Comment (salsakran) [inline src/metabase/automagic_dashboards/core.clj]: did all this go to the MBTL codebase or is it going to be re-implemented elsewhere?

File: resources/automagic_dashboards/table/TransactionTable.yaml (hunk 4 above)
Comment (sbelak) [reply resources/automagic_dashboards/table/TransactionTable.yaml]: These are meant more for things like "last 30 days" which are not really semantical breakdown of data, but I could see the argument for moving these to the data model definition as well.

File: src/metabase/automagic_dashboards/core.clj (hunk 5 above)
Comment (sbelak) [reply src/metabase/automagic_dashboards/core.clj]: Some went to MBTL, the rest is no longer needed. A lot of complexity was there to support varied shapes of data (via overloading and various ways of binding), which we no longer need to do as x-rays will just query the data model, and the data model decomposes the problem to: start from something known (& fixed), and than transform.

File: resources/data_models/orders.yaml (hunk 1 above)
Comment (sbelak) [reply resources/data_models/orders.yaml]: I'm not convinced. The `required_dimensions` and `optional_dimensions` are specifications about the data shape,  `breakout_dimensions`, `metrics`, & `segments` are what you can do with this. Could see having a better name though.

File: src/metabase/automagic_dashboards/core.clj (hunk 3 above)
Comment (sbelak) [reply src/metabase/automagic_dashboards/core.clj]: It's the YAML. We often use the term "heuristic" but that feels like it's leaking implementation into the vocabulary. Open to suggestions though

File: src/metabase/automagic_dashboards/core.clj (hunk 5 above)
Comment (salsakran) [reply src/metabase/automagic_dashboards/core.clj]: that's a good sign!

File: src/metabase/automagic_dashboards/core.clj (hunk 3 above)
Comment (salsakran) [reply src/metabase/automagic_dashboards/core.clj]: what's a heuristic?

Comment (salsakran) [reply src/metabase/automagic_dashboards/core.clj]: to be less glib, if you're loading a bunch of models, you should call those models.

if you're loading x-ray definitions, you should call them x-ray definitions.

Both "rules" and "heuristics" are too ambiguous

Comment (sbelak) [reply src/metabase/automagic_dashboards/core.clj]: /resources/automagic_dashboards/table/Transactions.yaml 🤷‍♂️

Comment (sbelak) [reply src/metabase/automagic_dashboards/core.clj]: Fair point. The terminology is a leftover from the first implementation, but yes, names need love/refactoring too

File: resources/data_models/orders.yaml (hunk 1 above)
Comment (sbelak) [reply resources/data_models/orders.yaml]: One learning lesson from writing a bunch of these yamls for xrays & etl is that you want the simplest shape possible both for writing as such, and to make validation more comprehensible

Comment (salsakran) [reply resources/data_models/orders.yaml]: I'm really squeamish about the weird separation of a fingerprint to match (eg, required + optional dimensions) + a seemingly super-specific interactive description of breakout dimensions. 

I.e. if we have a usage (say micro x-rays) where we want 3 breakout dimensions, but normally have 5, do we edit this? 

If this isn't an x-ray specific matter, then it gets even weirder for me. Are we really saying the model defines the only dimensions that can be broken out into? Are these meant to be a suggestions? Is it just whether they are a primary (vs secondary dimension)? 

Also, there will be tables where one of the fields is a metric (eg a rollup table).

Comment (sbelak) [reply resources/data_models/orders.yaml]: It's meant to be more than just for x-rays, but also for things like drilldowns etc. If you want less, you could just `(take 3)` and not meant as only these, but rather, if you don't know what, these are good. Having said that, I can see the argument, that in most cases, you can assume that for the most part these will equal to fields with `special_type` that is some derivative of `type/category`, and the rest of the cases are pretty context specific. 

As for the metric thing, that's already covered (but not explicit enough), as if you look at the MBTL, dimension can stand for a field (via it's `special_type` or `name`), metric, or expression. The reasoning being that one question's metric is a subsequent question's field[_literal].

**Comment by salsakran** [comment]
Regarding open questions re X-Rays:

`should there be some generic data model?`
Don't think so. There should be a default x-ray that takes arbitrary dimensions and tries to do something smart with them. The idea of a data model is that it models something. I don't want the idea that "an x-ray works on a data model" to cause us to abuse the definition of a "Data model".

`How do drilldown x-rays look?`
Good? IMO, there should be a "mini version" of each x-ray, and a more indepth version.
I'm honestly not a huge fan of the current drill-down behavior at the moment. However, I think the idea of exploring a dimension (eg, but clicking a header on a table) is a useful concept still. 
We should still have field x-rays imo that take metrics (etc) of the model they point to as well.


`Should we retain the notion of interestingness/card score? If so where/why?`
Yes, because there will often be too much to put on the screen. 
In the x-ray.
The model should also have some information on which dimensions or metrics are most important, but the final scoring, weighting and selection of specific charts is an x-ray concern.

**Comment by salsakran** [comment]
`how to model relations !`
Via the type system? =) What do you specifically mean here?

`grouping would be nice`
In general grouping is 👍 

`are string names for dimensions chill?`
What's the alternative?

File: src/metabase/models/data_model.clj [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  (best-fit (concat (etl/candidates table)
+                    (filter (partial satisfies-requirements? table) @data-models))))
+
+(defn- instantiate-template
+  "Bind dimensions and fill out metrics & segmets (reusing/extracting the same logic as ETL)"
+  [template table]
+  ;; ....
+  {:name        "Orders"
+   ;; Entity is either a table or a card (produced by ETL)
+   :entity      entity
```
Comment (salsakran) [inline src/metabase/models/data_model.clj]: please note what "entity" is here

**Comment by sbelak** [comment]
Current state of affairs:

Done:
* Transforms
* DomainSpec (language, parser, model, binding)

Next steps:
0) Fire off some small fixes to other parts of MB [1/2 day]
1) Wire up DomainSpec to sync & transform so we have a roundtrip there [1/2 day]
2) Rewrite x-rays [1 day]

At this point we have feature parity with much less code, cleaner architecture, and ability to do non-trivial transformations (see: churn).

Next next steps:
* Figure out how to do groups (a la Looker) better
* Make x-rays smaller and more composable 
* Write more transforms & x-rays

**Comment by sbelak** [comment]
Closing -- this has served it's purpose.

**Pull request closed without merging**